src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs

```diff
@@ -247,76 +373,105 @@ public SNIHandle CreateConnectionHandle(object callbackObject, string fullServer
+        private SNITCPHandle CreateTcpHandle(string fullServerName, long timerExpire, object callbackObject, bool parallel, ref byte[] spnBuffer, bool isIntegratedSecurity)
         {
             // TCP Format: 
             // tcp:<host name>\<instance name>
             // tcp:<host name>,<TCP/IP port number>
 
-            if (string.IsNullOrWhiteSpace(fullServerName))
-            {
-                int defaultInstancePort = TryToGetDefaultInstancePort(DefaultHostname);
-                return new SNITCPHandle(DefaultHostname, (defaultInstancePort > 0 ? defaultInstancePort : DefaultSqlServerPort), timerExpire, callbackObject, parallel);
-            }
-
-            string[] serverNamePartsByComma = fullServerName.Split(CommaSeparator);
-            string[] serverNamePartsByBackSlash = fullServerName.Split(BackSlashSeparator);
+            string hostName = null;
+            int port = -1;
+            Exception exception = null;
 
-            if (serverNamePartsByComma.Length < 2 && serverNamePartsByBackSlash.Length < 2)
+            if (string.IsNullOrWhiteSpace(fullServerName)) // when fullServerName is empty
             {
-                return new SNITCPHandle(fullServerName, DefaultSqlServerPort, timerExpire, callbackObject, parallel);
+                hostName = DefaultHostName;
+                port = DefaultSqlServerPort;
             }
-            else if (serverNamePartsByComma.Length == 2 && serverNamePartsByBackSlash.Length < 2)
+            else
             {
-                string hostName = serverNamePartsByComma[0];
-                int portNumber = -1;
-                try
+                string[] serverNamePartsByComma = fullServerName.Split(CommaSeparator);
+                string[] serverNamePartsByBackSlash = fullServerName.Split(BackSlashSeparator);
+
+                // when no port or instance name provided
+                if (serverNamePartsByComma.Length < 2 && serverNamePartsByBackSlash.Length < 2)
                 {
-                    portNumber = ushort.Parse(serverNamePartsByComma[1]);
-                    return new SNITCPHandle(hostName, portNumber, timerExpire, callbackObject, parallel);
+                    hostName = fullServerName;
+                    port = DefaultSqlServerPort;
                 }
-                catch (Exception e)
+                // when port is provided, and no instance name
+                else if (serverNamePartsByComma.Length == 2 && serverNamePartsByBackSlash.Length < 2)
                 {
-                    SNILoadHandle.SingletonInstance.LastError = new SNIError(SNIProviders.TCP_PROV, SNICommon.InvalidConnStringError, e);
-                    return null;
+                    hostName = serverNamePartsByComma[0];
+                    string portString = serverNamePartsByComma[1];
+                    try
+                    {
+                        port = ushort.Parse(portString);
```

Comment:
Our earlier approach was TryParse(), but we want to keep the reason why the parsing is failed, and send the error message to SNIError object.